net: conn_mgr: Cleanup & Connectivity Counting

glarsennordic opened the pull request:
Key features:
- moved from subsys/net/lib/conn_mgr to just subsys/net/conn_mgr
- event logic and flag names simplified
- thread-safety added
- basic filtering of ifaces and L2s added
- NET_EVENT_L4_CONNECTED/NET_EVENT_L4_DISCONNECTED now only fired when the first iface gains connectivity, and when the last iface loses connectivity, respectively

glarsennordic committed:
net: conn_mgr: Add thread-safety

Current implementation does not prevent the net_mgmt thread from
preempting during state change handling and generating more
state changes. Add a mutex to protect internal state of conn_mgr

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Reorganize files

Since conn_mgr is a subsystem rather than a library, relocate it
directly into subsys/net rather than subsys/net/lib/

Rename header files to better match their function.

Remove net_ prefix from conn_mgr types, API, and files, since it is
unnecessary.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: refactor event logic

In preparation for altering the conditions under which conn_mgr
fires NET_EVENT_L4_CONNECTED and NET_EVENT_L4_DISCONNECTED,
refactor the event logic without changing its behavior.

These changes do not alter the behavior of conn_mgr. They merely change
how that behavior is expressed so that it is easier to understand, and so
that upcoming changes are easier to implement.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Rename flags

Rename internal flags for clarity

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Fire events only on connectivity gain/loss

Currently, conn_mgr fires L4_CONNECTED and L4_DISCONNECTED events for
every individual iface that gains or loses connectivity.

After this commit, these events are only fired when the first iface gains
connectivity, and the last iface loses connectivity, respectively.

This is so that applications can rely on conn_mgr as a simple way to
track whether connectivity is available.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Ignore ifaces and L2s

Adds a flag to allow ifaces and L2s to be ignored.

This is so that applictions have better control over what ifaces they
want connectivity management for.

Enabling the flag forces conn_mgr to consider the iface
unready.

Also increase NET_CONNECTION_MANAGER_PRIORITY default value from 0 to 1
so that applications can register SYS_INIT callbacks that fire just
before conn_mgr initializes. This allows ignored ifaces to be configured
before conn_mgr starts, if needed.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented:
Currently a draft:

- Based on https://github.com/zephyrproject-rtos/zephyr/pull/55946, so ignore the first two commits
- Supercedes https://github.com/zephyrproject-rtos/zephyr/pull/54838
- Currently missing a test suite

glarsennordic committed:
net: conn_mgr: Add test suite

Create test suite covering all major features of conn_mgr

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: relocate if_get_binding

Relocate conn_mgr_if_get_binding to conn_mgr_private.h so that it
can be used by test suites.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic committed:
net: conn_mgr: Simplify DAD handling

The current method for handling DAD directly tracks a DAD state for each
iface.

This does not reflect the fact that each individual IP has a separate
DAD state.

This also does not align with the logic in conn_mgr_initial_state, which
instead uses net_if_ipv6_get_global_addr(NET_ADDR_PREFERRED, ...).

Thus, the current approach falsely reports IPv6 ready when DAD adds
the link_addr, and also fails to report IPv6 ready if DAD is disabled.

This commit removes DAD tracking completely, and relies solely on
net_if_ipv6_get_global_addr(NET_ADDR_PREFERRED, ...), fixing both
problems, and simplifying the event tracking.

For consistency, slight modifications to IPv4 handling as well.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented on tests/net/conn_mgr/src/main.c:
```diff
@@ -0,0 +1,921 @@
+	 * to execute it only when DAD is disabled, since we don't expect DAD to affect any
+	 * behaviors beyond the IP addition/removal testing it defeats.
+	 *
+	 * Compatibility of conn_mgr with DAD is ensured by cycle_ignored_iface succeeding both
+	 * with and without DAD (the suite is configured to run twice, once with each setting).
```
Thinking on it, this suite currently lacks a test explicitly requiring DAD to succeed as part of the test. I will write one.

glarsennordic commented on tests/net/conn_mgr/src/main.c:
```diff
@@ -0,0 +1,921 @@
+
+		/* Add IPv6 */
+		net_if_ipv6_addr_add(iface, &test_ipv6_a, NET_ADDR_MANUAL, 0);
+		k_sleep(DAD_WAIT_TIME);
```
DAD sleep no longer makes sense for this specific routine

wbober commented on include/zephyr/net/conn_mgr.h:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @param iface - iface to be ignored.
+ */
+void net_conn_mgr_ignore_iface(struct net_if *iface);
```
I think `*_watch` and `*_unwatch` could also work.

Should we explain what is the implication of ignoring an interface or it's just too obvious?

rlubos replied:
I don't really have a strong preference, but `net_conn_mgr_monitor_iface` could also be an option (with an extra bool parameter to enable/disable).

glarsennordic replied:
I don't have a strong preference on naming either, though I lean towards "ignore/unignore" simply because it suggests that ifaces are paid attention by default (which they are). @wbober Since Robert doesn't have a strong opinion, and you raised the issue originally, which of the three options would be your final choice?

I agree that the documentation for these functions needs to be improved. I will make sure to do that

glarsennordic replied:
I've improved the specificity of the documentation, although the behavior described technically requires a later commit to be included. I feel that is a lesser evil than rewriting the docs in the commit that changes the behavior. I will include a note about it in the commit message.

wbober commented on include/zephyr/net/conn_mgr.h:
```diff
@@ -0,0 +1,76 @@
+ * @brief Resend either NET_L4_CONNECTED or NET_L4_DISCONNECTED depending on whether connectivity
+ * is currently available.
+ */
+void net_conn_mgr_resend_status(void);
```
From the API user perspective, what is the purpose of that function?

glarsennordic replied:
It is a legacy feature from the previous version of `conn_mgr`. It appears to be provided so that applications that set up their event listeners after conn_mgr initialized can manually request the latest connectivity state be resent.

I think it might be worth it to eventually phase this function out, but I consider that task beyond the scope of these modifications

wbober commented on samples/net/sockets/echo_server/src/tunnel.c:
```diff
@@ -159,3 +159,13 @@ int init_tunnel(void)
+}
+
+BUILD_ASSERT(CONFIG_NET_CONNECTION_MANAGER_PRIORITY > 0);
+SYS_INIT(ignore_tunnel, APPLICATION, CONFIG_NET_CONNECTION_MANAGER_PRIORITY - 1);
```
I find this part a bit unfortunate. This is because you need to explicitly know to add this for a given interface type. How do you, as a developer, will know that it's needed?

Perhaps a method with an interface or L2 flag, which indicates if an interface is suitable for watching by `conn_mgr`, would make sense. This could be added in a subsequent PR. This way we shift the responsibility from the developer to integration author.

glarsennordic replied:
I'm not sure the responsibility should be with the integration provider; I can imagine a scenario where, for instance, where a connectivity-capable iface might be occasionally available, but the application developer does not want to rely on it. I also think that whether an iface is suitable for monitoring depends largely on context, and that context is defined in the application domain.

But I think I agree with you partially, that `net_conn_mgr_ignore_l2` is not really a great way for applications to control which ifaces they receive. The L2 type is often very disconnected from the actual nature of the iface. What we really need (and have discussed in other PRs) is some kind of (possibly extensible) tagging system for ifaces that is independent of the L2, and somehow communicates the underlying nature of each iface. Then we could allow applications to ignore/unignore by tag. This would also fix the responsibility issue by making it the responsibility of the integration author to accurately tag the ifaces they provide, while still allowing the application developer full control over what tags they are interested in.

But the conclusion we have come to in those discussions is that such a tagging system is beyond the scope of this initial implementation.

I mainly included `net_conn_mgr_ignore_l2` specifically because of this sample -- it relies on filtering by L2 type, and that is not possible with the new event behaviors. If we had preserved the old event behavior and added new events with the new behavior, this would not be necessary -- but that was the approach taken in https://github.com/zephyrproject-rtos/zephyr/pull/53118, and that approach did not stand up to review

rlubos replied:
I start to wonder, is `SYS_INIT` configuration even needed here? IPIP interfaces have `NET_IF_NO_AUTO_START` set, so they won't start at boot, and we have `init_tunnel()` here which configures IP addresses, so this interface should not be considered "ready" before that. Especially, that the carrier state of an virtual interface is controlled by the operational UP status of the underlying interface - so the virtual iface should not be set to oper UP before the real one.

So I think the logic could be simplified a bit, and the interface could be ignored simply in `init_tunnel()`.

glarsennordic replied:
Ah! Good point! I will switch to the approach you suggest

rlubos commented on subsys/net/conn_mgr/conn_mgr.c:
```diff
@@ -31,118 +31,126 @@ K_SEM_DEFINE(conn_mgr_event_signal, 1, 1);
 /* Used to protect conn_mgr state */
 K_MUTEX_DEFINE(conn_mgr_lock);
 
-static enum net_conn_mgr_state conn_mgr_iface_status(int index)
```
"Also rename a few flags." no longer seems to apply to commit 003a420e804be12170241674ab8a351f2de42492?

glarsennordic replied:
Good catch! You're right, I chose to do that in a separate commit to help make reviewing easier. I forgot I had mentioned it in the commit message

rlubos commented on subsys/net/conn_mgr/conn_mgr.c:
```diff
@@ -31,118 +31,126 @@ K_SEM_DEFINE(conn_mgr_event_signal, 1, 1);
+static void conn_mgr_set_ready(int idx, bool readiness)
+{
+	/* Clear and then update the L4-readiness bit */
+	iface_states[idx] &= ~CMGR_IF_READY;
```
nit: This commit could already use proper flags, instead of having them fixed in the next one.

glarsennordic replied:
I'm not 100% sure which commit you commented on here, so just to clarify, you are talking about https://github.com/zephyrproject-rtos/zephyr/pull/56549/commits/02c02bb5de16a004c18b2c62ad2d8683faffc7a4 right?

glarsennordic replied:
Whoops. I've had some troubles with propagating my event renames. Not sure how those changes keep ending up in later commits 🤦. Thanks for catching that, I'll get it corrected.

rlubos commented on include/zephyr/net/conn_mgr.h:
```diff
@@ -13,11 +13,59 @@ extern "C" {
+ * @brief Check whether the provided iface is currently ignored.
+ *
+ * @param iface - The iface to check.
+ * @return true if the iface is being ignored by conn_mgr.
+ * @return false otherwise.
```
If you document individual return values, `@retval` should be used instead.

glarsennordic replied:
Same with this

rlubos commented on samples/net/sockets/echo_server/src/echo-server.c:
```diff
@@ -84,6 +84,9 @@ static void stop_udp_and_tcp(void)
 static void event_handler(struct net_mgmt_event_callback *cb,
 			  uint32_t mgmt_event, struct net_if *iface)
 {
+	ARG_UNUSED(iface);
```
The changes in `echo_server` sample seem a bit unrelated to other work in this commit, perhaps they merit a separate commit?

glarsennordic replied:
I suppose these changes could be moved to a commit between this one and the commit before it -- But I included them in this commit since this is technically the commit that causes these changes to be required. Although in reality, the changes would work and are justifiable as early as hyr/pull/56549/commits/b8cf864359676657057323c35299cb533dbfdbc8

I'm pretty neutral, so final decision is up to you

rlubos replied:
Ah, after having a second look I now understand why are the changes related. I think that it could either be extracted into a separate commit after b8cf864, or, at very least, you could elaborate in the commit message on why are the changes needed.

glarsennordic replied:
I'll do both, extract as a separate commit, and I'll try to update all the commit messages to include a bit more why, as per @jukkar 's very good point here: https://github.com/zephyrproject-rtos/zephyr/pull/57057#discussion_r1172540424

glarsennordic committed:
samples: net: sockets: echo_server: Use conn_mgr ignore_iface

conn_mgr now allows applications to request that ifaces be ignored.

This commit changes the echo_server sample to use this feature to
filter out the tunnel iface, instead of manually filtering events
generated by it.

This will allow the echo_server sample to remain compatible with
planned changes to conn_mgr, namely a switch from events being
fired for every individual iface that goes up or down, to only
being fired when connectivity is gained or completely lost.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented on subsys/net/conn_mgr/conn_mgr_private.h:
```diff
@@ -49,4 +51,27 @@ enum net_conn_mgr_state {
+ * @return struct conn_mgr_conn_binding* Pointer to the retrieved binding struct if it exists,
+ *	   NULL otherwise.
+ */
+
```
Oops. I missed this before force-pushing. I'll nab it on the next revision.

rlubos reviewed (dismissed):
Thanks, LGTM

glarsennordic commented on include/zephyr/shell/shell_mqtt.h:
```diff
@@ -12,7 +12,7 @@
 #include <zephyr/net/socket.h>
```
It's more of a subjective decision, none of us felt extremely strongly about it, but the placement feels more natural as opposed to it being in net/lib, since it is not really a library but a subsystem. I didn't include that reasoning in the commit since it is more subjective, especially with how Zephyr is organized in general. There isn't, to my knowledge, an objective argument for one location or another

glarsennordic commented:
@jukkar I am having troubles running the tunneling overlay for echo_server, could you please advise? I see you are the one who added this overlay

I am able to run and ping the sample itself successfully in QEMU, but adding the tunnel overlay causes it to not get past boot.

I suspect it might be that I haven't configured `NET_SAMPLE_TUNNEL_PEER_ADDR` and `NET_SAMPLE_TUNNEL_MY_ADDR`, but I'm not actually sure what these should be set to. I might be able to figure that out, but I figure it makes sense to just, you know, ask the original author since you're right here 😅

jukkar commented:
> I suspect it might be that I haven't configured NET_SAMPLE_TUNNEL_PEER_ADDR and NET_SAMPLE_TUNNEL_MY_ADDR, but I'm not actually sure what these should be set to. I might be able to figure that out, but I figure it makes sense to just, you know, ask the original author since you're right here

Sorry, I missed your question earlier. Not sure if the issue is still valid, but regarding the tunneling support, the capture API is perhaps the best example how it is being used. See for example `subsys/net/lib/capture` for the API and `subsys/net/ip/net_shell.c` (the `net capture` cmd) how the various parameters are set and what they mean. The `samples/net/capture` sample will enable the capture shell support if you want to try it live.

jukkar commented on subsys/net/conn_mgr/events_handler.c:
```diff
@@ -77,25 +77,18 @@ static void conn_mgr_ipv6_events_handler(struct net_mgmt_event_callback *cb,
 	k_mutex_lock(&conn_mgr_lock, K_FOREVER);
 
 	switch (NET_MGMT_GET_COMMAND(mgmt_event)) {
+	case NET_EVENT_IPV6_CMD_DAD_SUCCEED: /* Intentional fallthrough */
```
Please use the `__fallthrough;` instead of the fallthrough comment.

jukkar commented on subsys/net/conn_mgr/events_handler.c:
```diff
@@ -77,25 +77,18 @@ static void conn_mgr_ipv6_events_handler(struct net_mgmt_event_callback *cb,
-	case NET_EVENT_IPV6_CMD_DAD_FAILED:
-		if (net_if_ipv6_get_global_addr(NET_ADDR_PREFERRED, &iface)) {
-			break;
+	case NET_EVENT_IPV6_CMD_DAD_FAILED:  /* Intentional fallthrough */
```
same here

glarsennordic commented:
@jukkar I took a look at those samples, and they were helpful to some extent, but an actual test procedure is still unclear to me.

I have tried the following:

Terminal 1, in net-tools: `sudo ./net-setup.sh -c zeth-tunnel.conf`
Terminal 2, in net-tools: `./loop-socat.sh`
Terminal 3, in net-tools: `sudo ./loop-slip-tap.sh`
Terminal 4, in echo-server sample: `west build -p always -b qemu_x86 samples/net/sockets/echo_server -- -DCONF_FILE=overlay-tunnel.conf` followed by `west build -t run`

I derived this by trying to combine the instructions from these two sample pages:
https://docs.zephyrproject.org/3.2.0/samples/net/capture/README.html (per your recommendation)
https://docs.zephyrproject.org/2.7.0/guides/networking/qemu_setup.html (standard instructions for this sample)

Terminal 4 gets stuck after boot.

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build zephyr-v3.2.0-7691-ga01d40f6049a ***
```

This is true whether or not I use the latest from this PR, or 100 commits before it, suggesting that the issue is not with this PR.

My theory about improperly configured NET_SAMPLE_TUNNEL_PEER_ADDR and NET_SAMPLE_TUNNEL_MY_ADDR was wrong -- the overlay configures the exact values recommended by the files you pointed me to.

Do you have any recommendations on a proper test procedure for this sample with the tunnel overlay?

glarsennordic committed:
net: conn_mgr: connect/disconnect adjustments

Rather than raise error on connect if iface is down, just take the iface
up.

Rather than raise error on disconnect if iface is down, just ignore the
disconnection request.

Adjust tests and documentation accordingly.

This should make the API easier to use for applications and is more in
line with the original L2 agnostic connectivity RFC.

Also replace some instances of net_if_flag_is_set with
net_if_is_admin_up, which is cleaner and identical in function.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

rlubos commented:
Given that tunnel problem is reproducible on main, I don't think it shoudld block this PR. Please open a bug report regarding this sample configuration, I'll give it a closer look.

jukkar commented on include/zephyr/net/conn_mgr.h:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @param iface - iface to no longer ignore.
+ */
+void conn_mgr_unignore_iface(struct net_if *iface);
```
This unignore word looks weird. I was not able to find it in dictionaries so it looks like it is not proper English. Dunno what we should have here, Perhaps `conn_mgr_start_ignore_iface()` and `conn_mgr_stop_ignore_iface()` although they are a bit longish.

glarsennordic replied:
What about `conn_mgr_ignore_iface` and `conn_mgr_watch_iface`? This would align better with the documentation I wrote anyways. @rlubos @wbober You also brought up the naming of these two functions, would this naming be satisfying to you as well?

glarsennordic replied:
I've gone ahead and force-pushed `ignore/unignore->ignore/watch` on the assumption/suspicion that everyone involved will either not care or prefer it. Let me know if this is satisfactory

jukkar replied:
Ignore/watch naming LGTM.

rlubos replied:
Sounds fine to me

jukkar commented on subsys/net/conn_mgr/conn_mgr.c:
```diff
@@ -0,0 +1,336 @@
+ */
+static struct net_if *conn_mgr_get_if_by_index(int index)
+{
+	return net_if_get_by_index(index + 1);
```
Why +1 here? It looks like you have now separate "index" space between this API and the net_if API.

jukkar replied:
Some background for the index numbering. Originally network interface numbering started from 0, but then it was decided that index 0 would indicate all the network interfaces so similar was as in various Linux APIs that deal with network interfaces numbering. This means that net_if internally does +1 and -1 when figuring out the actual interface in the interface array.
So even if the index here is only used internally, I would suggest that we follow the index numbering style done in net_if. It is just quite confusing otherwise for the future developer who is looking the code and trying to understand why the index numbering differs. Also this could avoid possible hard to find bugs.

glarsennordic replied:
I have now done as you suggest, Robert! @jukkar let me know if this change is satisfactory

glarsennordic commented:
If there are no further concerns, would you be willing to approve, @jukkar @rlubos ?

jukkar commented on subsys/net/conn_mgr/conn_mgr.c:
```diff
@@ -0,0 +1,339 @@
+/**
+ * @brief Retrieves pointer to an iface by the index that corresponds to it in iface_states
+ *
+ * @param index - The index in iface_states to find the corresponding iface for.
```
I would have thought that the index here is the interface index number, and not the iface[index]. This is what I mean that using the same name for this value is super confusing. Especially if the documentation does not clearly indicate what the value is like here. The iface_states is documented but imho documentation here would need some enhancements.

rlubos replied:
Isn't the following part explicit enough that the indexing refers to `iface_states` array?
> by the index that corresponds to it in **iface_states**

I can't really see an alternative for this function. And for details on what `iface_states` is, it has it's own documentation, I'm not sure if repeating the information here is beneficial.

jukkar commented on subsys/net/conn_mgr/conn_mgr.c:
```diff
@@ -0,0 +1,339 @@
+#endif
+
+/* Internal state array tracking readiness, flags, and other state information for all available
+ * ifaces. Note that indexing starts at 0, whereas Zephyr iface indices start at 1.
```
I am just wondering why the indexing cannot start from 1 for the iface_states too. In the getter/setter function you would then just do iface_states[index - 1]. Requires of course that one does not access the iface_states directly but via a function.

rlubos replied:
IMO it could then can confusing in other place, where we simply iterate over the array - it be needed then not to use C-style array indexing (from 0) but start from 1 and use getter/setter. Perhaps it's just a matter of preference, but I don't think the current approach is that confusing (we no longer store the modified index in `last_iface_down/up`).

jukkar replied:
Fair enough, if you think this is clear enough I can certainly live with this.
